src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,320 @@
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.itervalues():
+      heartbeat_requests.extend(cbsd_object_item.constructHeartbeatRequestForAllActiveGrants())
+
+    if len(heartbeat_requests):
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      # Perform heartbeat requests.
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_request, heartbeat_response in zip(heartbeat_requests, heartbeat_responses):
+        cbsd_object = self.cbsd_objects[heartbeat_request['cbsdId']]
+        grant_object = cbsd_object.grant_objects[heartbeat_request['grantId']]
+
+        # Update grant state based on the heartbeat response.
+        self._mapResponseCodeToGrantState(heartbeat_response['response']['responseCode'],
+                                         grant_object)
+
+      return heartbeat_requests, heartbeat_responses
+
+  def _mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state."""
+    if heartbeat_response_code == ResponseCodes.SUCCESS.value:
+      grant_object.authorized_in_last_heartbeat = True
+    elif heartbeat_response_code == ResponseCodes.TERMINATED_GRANT.value:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == ResponseCodes.SUSPENDED_GRANT.value:
+      grant_object.authorized_in_last_heartbeat = False
+    else:
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
+      raise Exception('Unknown code received in Heartbeat Response')
+
+  def _constructGrantRequest(self,heartbeat_response):
+    """Construct unwrapped version of grant request."""
+    grant_request = {
+      'cbsdId': heartbeat_response['cbsdId'],
+      'operationParam': {
+        'maxEirp': heartbeat_response['operationParam']['maxEirp'],
+        'operationFrequencyRange': {
+          'lowFrequency': heartbeat_response['operationParam']['operationFrequencyRange']['lowFrequency'],
+          'highFrequency': heartbeat_response['operationParam']['operationFrequencyRange']['highFrequency']
+        }
+      }
+    }
+    return grant_request
+
+  def performHeartbeatAndUpdateGrants(self):
+    """Does heartbeat request for all active grants and updates the grants.
+
+    The following steps are performed by this function
+      1. Send heartbeat request on all grant of CBSD at once
+      2. If the heartbeat response contains operationParam then those grants
+         are relinquished.
+      3. New grant requests with the suggested operationParam values in the
+         heartbeat response is sent. Failures are ignored.
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
```

Comment:
heartbeat_requests is assigned on the next line, remove this line.

Reply:
Removed   as per suggestion.